src/Shared/Components/ComponentDescriptorSerializer.Deserialize.cs

```diff
@@ -0,0 +1,215 @@
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Runtime.InteropServices;
+using System.Security.Cryptography;
+using System.Text.Json;
+using Microsoft.AspNetCore.DataProtection;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNetCore.Components
+{
+    // ComponentDescriptorSerializer is an abstraction shared by both MVC and Blazor
+    // This file holds the definition used by Blazor to deserialize descriptors.
+    internal partial class ComponentDescriptorSerializer
+    {
+        public ComponentDescriptorSerializer(
+            IDataProtectionProvider dataProtectionProvider,
+            ILogger<ComponentDescriptorSerializer> logger,
+            RootComponentTypeCache rootComponentTypeCache)
+        {
+            DataProtector = dataProtectionProvider.CreateProtector(DataProtectionProviderPurpose);
+            Logger = logger;
+            RootComponentTypeCache = rootComponentTypeCache;
+        }
+
+        public ILogger<ComponentDescriptorSerializer> Logger { get; }
+
+        public RootComponentTypeCache RootComponentTypeCache { get; }
+
+        public bool TryDeserializeComponentDescriptorCollection(string serializedComponentRecords, out List<ComponentDescriptor> descriptors)
+        {
+            var markers = JsonSerializer.Deserialize<IEnumerable<ComponentMarker>>(serializedComponentRecords, _jsonSerializationOptions);
+            descriptors = new List<ComponentDescriptor>();
+            int lastSequence = -1;
+
+            var previousInstance = new ComponentDescriptorInstance();
+            foreach (var marker in markers)
+            {
+                if (marker.Type != ComponentMarker.ServerMarkerType)
+                {
+                    Log.InvalidMarkerType(Logger, marker.Type);
+                    descriptors.Clear();
+                    return false;
+                }
+
+                if (marker.Sequence == null)
+                {
+                    Log.MissingMarkerSequence(Logger);
+                    descriptors.Clear();
+                    return false;
+                }
+
+                if (marker.Descriptor == null)
+                {
+                    Log.MissingMarkerDescriptor(Logger);
+                    descriptors.Clear();
+                    return false;
+                }
+
+                var (descriptor, instance) = DeserializeComponentDescriptor(marker);
+                if (descriptor == null)
+                {
+                    // We failed to deserialize the component descriptor for some reason.
+                    descriptors.Clear();
+                    return false;
+                }
+
+                // We force our client to send the descriptors in order so that we do minimal work.
+                if (lastSequence != -1 && lastSequence != instance.Sequence - 1)
```

Comment:
I'd really like to understand if this is important, or even safe.

Reply:
Having a sequence number allows the server to uniquely identify each rendered component. This prevents the client from simply sending the same descriptor multiple times to cause the server to render the component multiple times on a single connection. 

For example, if there was no component sequence, I could do as follows:
* Send a request to the main page to generate  a valid descriptor.
* Create an array with as many copies of that descriptor as can fit in 32K.
* Open a connection to the server.
* Send that payload descriptor to the server and force it to render as many components as there are in the array.

The sequence number prevents a client from doing that, as a client can only render as many descriptors as the server generated in the first place. To put it in different words, this prevents individual descriptor replay attacks.

If you were to trick the server into generating a large amount of components in the first place (by an application introduced bug), you could ask the server to render at most that number of components, but no more. We (Steve and I) think that that is a legitimate answer and that it's a bug in the app. It can happen if you are not careful and for example render a root component for each element in a list and don't establish a limit, but in practical terms we don't expect this to be a common case. (Specially based on the fact that you can't currently render parameters).

What we expect is that you tipically render a few components in different parts of a page and this mechanism automatically limits the number of components to what the server rendered. As a last resort/fallback mechanism you are going to be limited by the signalr message size, but that's a limit that can be lifted, so we want something smarter and more adjustable like this.

Less important, the sequence number also allows the server to bail early for invalid data. If you include a descriptor that can't be unprotected or is out of sequence we detect it as soon as possible without having to potentially parse and unprotect the entire list of descriptors.

Reply:
Discussed offline with @GrabYourPitchforks